hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightEndpoint.java

```diff
@@ -0,0 +1,412 @@
+          // skip the prev key if prev key is present
+          if (skipPrevKey && key.equals(prevKeyPrefix)) {
+            continue;
+          }
+          KeyEntityInfo keyEntityInfo = new KeyEntityInfo();
+          keyEntityInfo.setKey(key);
+          keyEntityInfo.setPath(omKeyInfo.getKeyName());
+          keyEntityInfo.setInStateSince(omKeyInfo.getCreationTime());
+          keyEntityInfo.setSize(omKeyInfo.getDataSize());
+          keyEntityInfo.setReplicatedSize(omKeyInfo.getReplicatedSize());
+          keyEntityInfo.setReplicationConfig(omKeyInfo.getReplicationConfig());
+          openKeyInsightInfo.setUnreplicatedTotal(
+              openKeyInsightInfo.getUnreplicatedTotal() +
+                  keyEntityInfo.getSize());
+          openKeyInsightInfo.setReplicatedTotal(
+              openKeyInsightInfo.getReplicatedTotal() +
+                  keyEntityInfo.getReplicatedSize());
+          boolean added =
+              isLegacyBucketLayout ? nonFSOKeyInfoList.add(keyEntityInfo) :
+                  fsoKeyInfoList.add(keyEntityInfo);
+          if ((nonFSOKeyInfoList.size() + fsoKeyInfoList.size()) == limit) {
+            recordsFetchedLimitReached = true;
+            break;
+          }
+        }
+      } catch (IOException ex) {
+        throw new WebApplicationException(ex,
+            Response.Status.INTERNAL_SERVER_ERROR);
+      } catch (IllegalArgumentException e) {
+        throw new WebApplicationException(e, Response.Status.BAD_REQUEST);
+      } catch (Exception ex) {
+        throw new WebApplicationException(ex,
+            Response.Status.INTERNAL_SERVER_ERROR);
+      }
+      if (recordsFetchedLimitReached) {
+        break;
+      }
+    }
+    return Response.ok(openKeyInsightInfo).build();
+  }
+
+  /** This method retrieves set of keys/files/dirs pending for deletion. */
+  @GET
+  @Path("pendingfordeletionkeyinfo")
+  public Response getDeletedKeyInfo(
+      @DefaultValue(DEFAULT_FETCH_COUNT) @QueryParam(RECON_QUERY_LIMIT)
+      int limit,
+      @DefaultValue(StringUtils.EMPTY) @QueryParam(RECON_QUERY_PREVKEY)
+      String prevKeyPrefix) {
+    KeyInsightInfoResp deletedKeyAndDirInsightInfo = new KeyInsightInfoResp();
+    KeyInsightInfoResp pendingForDeletionKeyInfo =
+        getPendingForDeletionKeyInfo(limit, prevKeyPrefix,
+            deletedKeyAndDirInsightInfo);
+    return Response.ok(getPendingForDeletionDirInfo(limit, prevKeyPrefix,
+        pendingForDeletionKeyInfo)).build();
+  }
+
+  private KeyInsightInfoResp getPendingForDeletionDirInfo(
+      int limit, String prevKeyPrefix,
+      KeyInsightInfoResp pendingForDeletionKeyInfo) {
+
+    List<KeyEntityInfo> deletedDirInfoList =
+        pendingForDeletionKeyInfo.getDeletedDirInfoList();
+
+    Table<String, OmKeyInfo> deletedDirTable =
+        omMetadataManager.getDeletedDirTable();
+    try (
+        TableIterator<String, ? extends Table.KeyValue<String, OmKeyInfo>>
+            keyIter = deletedDirTable.iterator()) {
```

Comment:
Close the `keyIter` here as well

Reply:
this iterator is opened in try with block and any exception thrown will not be a leak which is taken care by java. Standard way to use in many other places in Ozone code base.